File: backend/src/models/bookbox.model.ts
import mongoose from 'mongoose';

const bookSchema = new mongoose.Schema({
   isbn: { type: String },
   title: { type: String, required: true },
   authors: [String],
   description: String,
   coverImage: String,
   publisher: String,
   categories: [String],
   parutionYear: Number,
   pages: Number,
   dateAdded: { type: Date, default: Date.now }
});

const bookboxSchema = new mongoose.Schema({
   name: { type: String, required: true },
   owner: { type: String, required: true },
   image: String,
   longitude: { type: Number, required: true },
   latitude: { type: Number, required: true },
   boroughId: { type: String, required: true },
   booksCount: { type: Number, default: 0 },
   location: {
       type: { type: String, default: 'Point' },
       coordinates: [Number] // [longitude, latitude]
   },
   infoText: String,
   books: [bookSchema],
   isActive: { type: Boolean, default: true },
});

bookboxSchema.pre('save', function(next) {
   // Always sync booksCount with books array length
   this.booksCount = this.books.length;
   
   // Sync location coordinates
   this.location = {
       type: 'Point',
       coordinates: [this.longitude, this.latitude]
   };
   
   next();
});

// INDEXES FOR NESTED BOOKS SEARCHING
bookboxSchema.index({
   "books.title": "text",
   "books.authors": "text", 
   "books.categories": "text"
});

bookboxSchema.index({ "books.title": 1 });
bookboxSchema.index({ "books.authors": 1 });
bookboxSchema.index({ "books.parutionYear": 1 });
bookboxSchema.index({ "books.dateAdded": 1 });
bookboxSchema.index({ "books._id": 1 });

// BOOKBOX-LEVEL INDEXES
bookboxSchema.index({ owner: 1 });
bookboxSchema.index({ isActive: 1 });
bookboxSchema.index({ boroughId: 1 });
bookboxSchema.index({ booksCount: 1 });
bookboxSchema.index({ name: 1 });

// GEOSPATIAL INDEX FOR LOCATION QUERIES
bookboxSchema.index({ location: '2dsphere' });

// COMPOUND INDEXES FOR COMMON QUERY PATTERNS
// Public search (active bookboxes with sorting)
bookboxSchema.index({ isActive: 1, booksCount: -1 });
bookboxSchema.index({ isActive: 1, name: 1 });

// Owner management queries
bookboxSchema.index({ owner: 1, isActive: 1 });
bookboxSchema.index({ owner: 1, name: 1 });

// Borough-based searches
bookboxSchema.index({ boroughId: 1, isActive: 1 });
bookboxSchema.index({ boroughId: 1, booksCount: -1 });

// Text search with activity filter
bookboxSchema.index({ isActive: 1, name: "text", infoText: "text" });

const BookBox = mongoose.model('BookBox', bookboxSchema, "bookboxes");

export default BookBox;

File: app/lib/models/bookbox_model.dart
import 'book_model.dart';

class BookBox {
  final String id;
  final String name;
  final String? image; // Optional image URL
  final double longitude;
  final double latitude;
  final String? infoText;
  final String boroughId;
  final List<Book> books;
  final String owner;
  final bool isActive;

  BookBox({
    required this.id,
    required this.name,
    this.image,
    required this.longitude,
    required this.latitude,
    required this.boroughId,
    this.infoText,
    required this.books,
    required this.owner,
    required this.isActive,
  });

  factory BookBox.fromJson(Map<String, dynamic> json) {
    var booksList = json['books'] as List;
    List<Book> books = booksList.map((bookJson) => Book.fromJson(bookJson)).toList();

    return BookBox( 
      id: json['_id'],
      name: json['name'],
      image: json['image'],
      longitude: (json['longitude'] as num).toDouble(),
      latitude: (json['latitude'] as num).toDouble(),
      boroughId: json['boroughId'],
      infoText: json['infoText'],
      books: books,
      owner: json['owner'],
      isActive: json['isActive'],
    );
  }
}

class BookBoxWithDistance extends BookBox {
  final double? distance; // Distance from the user's location

  BookBoxWithDistance({
    required super.id,
    required super.name,
    super.image,
    required super.longitude,
    required super.latitude,
    required super.boroughId,
    super.infoText,
    required super.books,
    this.distance,
    required super.owner,
    required super.isActive,
  });
}

class ShortenedBookBox {
  final String id;
  final String name;
  final String? infoText; // Optional info text
  final String? image; // Optional image URL
  final double longitude;
  final double latitude;
  final String boroughId;
  final int booksCount;
  final String owner;
  final bool isActive;
  final double? distance; // Optional distance from the user's location 

  ShortenedBookBox({
    required this.id,
    required this.name,
    this.infoText,
    this.image,
    required this.longitude,
    required this.latitude,
    required this.boroughId,
    required this.booksCount,
    required this.owner,
    required this.isActive,
    this.distance,
  });

  factory ShortenedBookBox.fromJson(Map<String, dynamic> json) {
    return ShortenedBookBox(
      id: json['_id'],
      name: json['name'],
      infoText: json['infoText'],
      image: json['image'],
      longitude: (json['longitude'] as num).toDouble(),
      latitude: (json['latitude'] as num).toDouble(),
      booksCount: json['booksCount'],
      owner: json['owner'],
      isActive: json['isActive'],
      boroughId: json['boroughId'],
      distance: json['distance'] != null ? (json['distance'] as num).toDouble() : null, // Handle optional distance
    );
  }
}

